lib/client/db/postgres/repl/repl_test.go

```diff
@@ -0,0 +1,418 @@
+	dataTypeQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "data_type", []byte(dataTypeQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "data_type")), dataTypeQueryResult)
+}
+
+// TestQuery given some input lines, the REPL should execute the expected
+// query on the PostgreSQL test server.
+func TestQuery(t *testing.T) {
+	ctx := context.Background()
+	_, tc := StartWithServer(t, ctx, WithCustomQueries())
+
+	// Consume the REPL banner.
+	_ = readUntilNextLead(t, tc)
+
+	for name, tt := range map[string]struct {
+		lines         []string
+		expectedQuery string
+	}{
+		"query":                     {lines: []string{"SELECT 1;"}, expectedQuery: "SELECT 1;"},
+		"query multiple semicolons": {lines: []string{"SELECT 1; ;;"}, expectedQuery: "SELECT 1; ;;"},
+		"query multiple semicolons with trailing space": {lines: []string{"SELECT 1; ;;  "}, expectedQuery: "SELECT 1; ;;"},
+		"multiline query":                     {lines: []string{"SELECT", "1", ";"}, expectedQuery: "SELECT\r\n1\r\n;"},
+		"malformatted":                        {lines: []string{"SELECT err;"}, expectedQuery: "SELECT err;"},
+		"query with special characters":       {lines: []string{"SELECT 'special_chars_!@#$%^&*()';"}, expectedQuery: "SELECT 'special_chars_!@#$%^&*()';"},
+		"leading and trailing whitespace":     {lines: []string{"   SELECT 1;   "}, expectedQuery: "SELECT 1;"},
+		"multiline with excessive whitespace": {lines: []string{"   SELECT", "    1", "     ;"}, expectedQuery: "SELECT\r\n1\r\n;"},
+		// Commands should only be executed if they are at the beginning of the
+		// first line.
+		"with command in the middle":              {lines: []string{"SELECT \\d 1;"}, expectedQuery: "SELECT \\d 1;"},
+		"multiline with command in the middle":    {lines: []string{"SELECT", "\\d", ";"}, expectedQuery: "SELECT\r\n\\d\r\n;"},
+		"multiline with command in the last line": {lines: []string{"SELECT", "1", "\\d;"}, expectedQuery: "SELECT\r\n1\r\n\\d;"},
+	} {
+		t.Run(name, func(t *testing.T) {
+			for _, line := range tt.lines {
+				writeLine(t, tc, line)
+			}
+
+			select {
+			case query := <-tc.QueryChan():
+				require.Equal(t, tt.expectedQuery, query)
+			case <-time.After(time.Second):
+				require.Fail(t, "expected to receive query but got nothing")
+			}
+
+			// Always expect a query reply from the server.
+			_ = readUntilNextLead(t, tc)
+		})
+	}
+}
+
+func writeLine(t *testing.T, c *testCtx, line string) {
+	t.Helper()
+	data := []byte(line + lineBreak)
+
+	// When writing to the connection, the terminal emulator always writes back.
+	// If we don't consume those bytes, it will block the ReadLine call (as
+	// we're net.Pipe).
+	go func(conn net.Conn) {
+		buf := make([]byte, len(data))
+		for {
+			n, err := conn.Read(buf[0:])
+			if err != nil {
+				t.Logf("Error while terminal reply on write: %s", err)
+				break
+			}
+
+			if string(buf[:n]) == line+lineBreak {
+				break
+			}
+		}
```

Comment:
nit: is the loop necessary? if `string(buf[:n]) == line+lineBreak` fails, the loop will keep going?

Reply:
We need to consume the prompt from the terminal. When it writes the reply, the prompt is empty. But when we have a multiline command, it is the empty reply + a reply with lead spacing.

We could limit this loop to consume these prompt replies. I don't know if that would be easier to understand. Anyway, I'll add this info to the code.

Reply:
i see. my concern is currently `writeLine` does not guarantee that the goroutine will finish when the function finishes, so the goroutine may keep running until the connection is closed, for example, if `line+lineBreak ` never come

Reply:
If this goroutine is stuck by any means, when the Cleanup kicks in (and closes the connection), it should free this goroutine.